### ISSUE TITLE
Thoughts on attempt at trio and wsproto integration

### ISSUE THREAD

**Issue by merrellb** [python-trio/trio#124]
So I've taken my first stab at writing a trio integration, using wsproto (@njsmith I've seen your contributions over there so if you already have a better integration with trio do let me know :-)

https://github.com/merrellb/trio_libs

I am particularly curious about any thoughts on my cancellable scope factory (is there any easier way to shutdown a single Websocket gracefully?) as well as my socket and nursery scope management base class (hopefully I have those nested correctly).  Also wasn't sure if a shutdown task (awaiting a shutdown event) is best-practice.  Basically, any feedback is much appreciated :-)

One issue I am running into is KeyboardInterrupt handling and allowing a graceful shutdown.  I've solved the initial issue with the socket closing prematurely (thanks for responding so quickly to my ticket)  but it still seems to hang on occasion requiring a double-control-c.   The eventual tracebacks are inconsistent, making it a bit hard to unravel.

My apologies if a ticket is not the best way to solicit feedback but figured some other folks would be interested in seeing an attempt at some trio code.

**Comment by njsmith** [comment]
Hey, this is great! Two quick thoughts before I look at the actual code:
* Have you seen #1?
* Regarding the double-control-c: are you by chance on windows?

**Comment by merrellb** [comment]
- I am subscribed to the issue but totally forgot to pin my trio version.  TODO requirements.txt
- I am on the latest OSX (errr MacOs)

**Comment by merrellb** [comment]
A quick update.  I think I've managed to conquer the KeyboardInterrupt problems.  The main issue was that I couldn't figure out how to wrap the implicit await (for spawned tasks to finish) at the end of a nursery scope.  Catching KeyboardInterrupt from the spawns wasn't helpful so I eventually decided to catch it outside the nursery and then just use a new nursery to handle shutting down all the connections in parallel.
